Enhancements to package.py

iskunk opened the pull request:
This PR depends on changes from [an associated PR](https://github.com/ungoogled-software/ungoogled-chromium/pull/3034) over in the main u-c repo. This is, likewise, a byproduct of my work on #293.

* Add the architecture identifier to the generated package files, so that x64 and x86 builds can be distinguished:

  |    | Installer | Zip file |
  |----|-----------|----------|
  | Old names | `ungoogled-chromium_123.0.4567.89-1.2_installer.exe` | `ungoogled-chromium_123.0.4567.89-1.2_windows.zip` |
  | New names | `ungoogled-chromium_123.0.4567.89-1.2_installer_x64.exe` | `ungoogled-chromium_123.0.4567.89-1.2_windows_x64.zip` |

  The build architecture is obtained by examining the `out/Default/args.gn` file (not sure if there's a better way).

* Look for the `build/util/LASTCHANGE.committime` file in the source tree, which contains the official timestamp for the Chromium release. If this file is present, then read the (Unix epoch format) timestamp therein, and pass it as a new parameter to `filescfg.create_archive()` so that the archive file can be created reproducibly.

  (If there is no timestamp, then member file timestamps go into the archive as-is, same as now.)

* Instead of deleting e.g. `mini_installer.exe` so that it doesn't get packed into the `.zip` file, use the new `excluded_files` parameter to `filescfg.filescfg_generator()`. This allows the packaging script to run idempotently---you can run it multiple times, and it won't bug out due to missing files after the first run.

  (I was running this script repeatedly in development, and having to rebuild/restore the missing files each time was quite tedious)

tomasz1986 commented:
Maybe follow the same name scheme as already used in this repository for releases?

![image](https://github.com/user-attachments/assets/bb364b1e-cafa-431e-b1d5-1b9e853c8fee)

iskunk commented:
Ah, indeed. I've updated the `.format()` lines, and the PR text above, to place the architecture ID at the end of the filenames.

Nifury commented:
You might also need to remove these lines.
https://github.com/ungoogled-software/ungoogled-chromium-windows/blob/4d0cfaa9917b566fa589ff540f276150d0639cdf/.github/actions/stage/index.js#L45-L52

iskunk commented:
> You might also need to remove these lines.

Ah, so that's the bit that was adding the architecture tags before posting the files to a release. Guess that won't be needed anymore.

I've added that edit here (also fixing the file's lack of a final newline), though I'm unable to test it myself.

iskunk commented:
The necessary backend changes are now merged in the main repo. Ready and willing to answer any questions.

Nifury commented:
```Traceback (most recent call last):
  File "F:\ungoogled-chromium-windows\build.py", line 323, in <module>
    main()
  File "F:\ungoogled-chromium-windows\build.py", line 201, in main
    downloads.retrieve_downloads(download_info_win, downloads_cache, True, args.disable_ssl_verification)
  File "F:\ungoogled-chromium-windows\ungoogled-chromium\utils\downloads.py", line 290, in retrieve_downloads
    if components and not download_name in components:
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: argument of type 'bool' is not iterable```

Is this a Windows-specific issue, or does Python 3.12 change the operator precedence?

iskunk commented:
My apologies, I forgot to update `build.py` with the new arguments from the other PR. (I'm not using this script, as my cross build does not run on Windows.) I've added those in, as well as an extra arg that was missing from the calls to `downloads.unpack_downloads()` (namely `skip_unused`, which I've set to `False` for now).

I've rebased and updated this PR. Could you try it again?

Nifury reviewed (approved):
Works great, thank you!

teeminus commented:
Seems to work fine for the upcoming .100 release. Do you want to merge?

iskunk commented:
Thanks everyone! Next up is getting the cross build ready for a PR, now that the necessary script updates are in.
